internal/jujuapi/applicationoffers.go

```
@@ -178,11 +178,20 @@ func (r *controllerRoot) modifyOfferAccess(ctx context.Context, change jujuparam
 		if err := r.jimm.PermissionManager().GrantOfferAccess(ctx, r.user, change.OfferURL, ut, change.Access); err != nil {
 			return errors.E(op, err)
 		}
+		// We grant access on the controller because pre 3.6.6 controllers are consulting their database to check if
+		// user has access to the offer when getting consume details. This is a bug and will be fixed in the following
+		// releases of Juju.
+		if err := r.jimm.JujuManager().GrantOfferAccessOnController(ctx, r.user, ut, change.OfferURL, change.Access); err != nil {
+			return errors.E(op, err)
+		}
 		return nil
 	case jujuparams.RevokeOfferAccess:
 		if err := r.jimm.PermissionManager().RevokeOfferAccess(ctx, r.user, change.OfferURL, ut, change.Access); err != nil {
 			return errors.E(op, err)
 		}
+		if err := r.jimm.JujuManager().RevokeOfferAccessOnController(ctx, r.user, ut, change.OfferURL, change.Access); err != nil {
+			return errors.E(op, err)
+		}
```

thinking out loud.

In case of removal is better to remove it first from jimm and then from juju, because if it fails we don't have an issue, we won't display this offer, etc, etc.

But in case of granting the permission, would it make sense to do the opposite? Before we add it to juju and then to jimm.
So, if juju fails the user doesn't see it when it lists offers and trigger the grant command again and we have a chance to try and re-add it to jimm/juju?

---

done

---

If we have this new strategy we should ignore error like "duplicated grant", otherwise we add it to JIMM it errors saying it's duplicated and we don't fwd to juju.

---

i think the permission manager does not return an error if relation already exists.

---

Can you add a test for it?

---

done